File: list-reduction/josephus.rb
class Josephus
    def initialize(count,nth)
        @count = count
        @nth = nth
    end
    def shout()
        counter = 0
        people = (1..@count).to_a
        while people.size > 1
            new_people = Array.new
            for index in 0 ... people.size
                if (counter + index) % @nth != 0 then
                    new_people.push people[index]
                end
            end
            counter = (counter + people.size) % @nth
            new_people.size
            people = new_people
        end
        return people[0]
    end
end

j = Josephus.new(40,3)
puts j.shout()
TIMES=10
ITER=1000000
TIMES.times { |i|
    start = Time.now
    ITER.times { |k|
        j.shout()
    }
    ends = Time.now
    puts ((ends - start) * 1000000 / ITER).to_s()
}

File: list-reduction/josephus.erl
-module(josephus).
-export([benchmark/0]).
-compile(native).

shout(Count,Nth) ->
    shoutList(lists:seq(1,Count),Nth,1).

shoutList([Head|[]],_,_) -> Head;
shoutList(List,N,Counter) ->
    {FilteredList, NextCounter} = shoutSubList(List,N,Counter,[]),
    shoutList(FilteredList,N, NextCounter).

shoutSubList([],_,Counter,Acc) -> {lists:reverse(Acc),Counter};
shoutSubList([_Head|Tail], N, 1, Acc) -> shoutSubList(Tail, N, 2, Acc);
shoutSubList([Head|Tail], N, Counter,Acc) ->
    NextCounter = 
        if
            Counter =:= N -> 1;
            true -> Counter + 1
        end,
    shoutSubList(Tail,N, NextCounter, [Head | Acc]).

run_iterations(0) -> ok;
run_iterations(Iterations) ->
    shout(40,3),
    run_iterations(Iterations - 1).

run_times(_,0) -> ok;
run_times(Iterations, Times) ->
    Start = os:timestamp(),
    run_iterations(Iterations),
    End = os:timestamp(),
    io:format("~w~n",[timer:now_diff(End,Start) / Iterations]),
    run_times(Iterations,Times-1).

benchmark() ->
    Iter = 1000000,
    io:format("(list reduction)~n",[]),
    run_times(Iter,10).
